edge-hub/watchdog/linux/src/main.rs

```diff
@@ -0,0 +1,182 @@
+                name: name,
+                process: child,
+            };
+
+            while child_process.is_running() && !has_parent_started_shutdown.load(Ordering::Relaxed)
+            {
+                let poll_interval: Duration = Duration::new(PROCESS_POLL_INTERVAL_SECS, 0);
+                thread::sleep(poll_interval);
+            }
+
+            child_process.shutdown_if_running();
+
+            has_shutdown_listener.store(true, Ordering::Relaxed);
+        });
+
+        has_shutdown
+    }
+
+    fn shutdown_if_running(&mut self) {
+        if self.is_running() {
+            info!("Terminating {:?} process", self.name);
+            self.send_signal(Signal::SIGTERM);
+        } else {
+            info!("{:?} process has stopped", self.name)
+        }
+
+        self.wait_for_exit();
+
+        if self.is_running() {
+            info!("Killing {:?} process", self.name);
+            self.send_signal(Signal::SIGKILL);
+        }
+    }
+
+    fn send_signal(&mut self, signal: Signal) {
+        info!("Sending {:?} signal to {:?} process", signal, self.name);
+        if let Err(e) = signal::kill(Pid::from_raw(self.process.id() as i32), signal) {
+            error!("Failed to send signal to {:?} process. {:?}", self.name, e);
+        }
+    }
+
+    fn is_running(&mut self) -> bool {
+        match self.process.try_wait() {
+            Ok(status) => status.is_none(),
+            Err(e) => {
+                error!(
+                    "Error while polling {:?} process status. {:?}",
+                    self.name, e
+                );
+                false
+            }
+        }
+    }
+
+    fn wait_for_exit(&mut self) {
+        let mut elapsed_secs = 0;
+        while elapsed_secs < PROCESS_SHUTDOWN_TOLERANCE_SECS && self.is_running() {
+            let poll_interval: Duration = Duration::new(PROCESS_POLL_INTERVAL_SECS, 0);
+            thread::sleep(poll_interval);
+            elapsed_secs += PROCESS_POLL_INTERVAL_SECS;
+        }
+    }
+}
+
+fn main() -> Result<(), Error> {
+    init_logging();
+    info!("Starting watchdog");
+
+    let should_shutdown = match register_shutdown_listener() {
+        Ok(should_shutdown) => should_shutdown,
+        Err(e) => {
+            error!(
+                "Failed to register sigterm listener. Shutting down. {:?}",
+                e
+            );
+            exit(1);
+        }
+    };
+
+    let has_broker_shutdown = ChildProcess::run_child_process(
+        "MQTT Broker".to_string(),
+        Command::new("/usr/local/bin/mqttd").stdout(Stdio::inherit()),
+        Arc::clone(&should_shutdown),
+    );
+
+    let has_edgehub_shutdown = ChildProcess::run_child_process(
+        "Edge Hub".to_string(),
+        Command::new("dotnet")
+            .arg("/app/Microsoft.Azure.Devices.Edge.Hub.Service.dll")
+            .stdout(Stdio::inherit()),
+        Arc::clone(&should_shutdown),
+    );
+
+    while !should_shutdown.load(Ordering::Relaxed) {
```

Comment:
So we are sharing the watchdog's sigterm/sigint listener bool across threads, so you are right that there is no need for ```while !should_shutdown.load(Ordering::Relaxed)```. 

But ```has_broker_shutdown``` and ```has_edgehub_shutdown``` are not shared between threads. Currently, if either broker / edgehub stop then this while loop exits and we use the ```should_shutdown``` shared bool to signal the child threads to shutdown. How would this work with join()? If a join() call is waiting then it needs to be able to determine when the other child process shut down, so that it can shut itself down. Should I use channels to communicate across child threads directly? Not sure how this would work, but it would solve this problem right?

I will look into channels.

Reply:
I don't think you need channels here. Let's discuss it offline?